lib/Sema/CSDiag.cpp

```diff
@@ -4635,14 +4635,27 @@ static bool diagnoseImplicitSelfErrors(Expr *fnExpr, Expr *argExpr,
   // If argument wasn't properly type-checked, let's retry without changing AST.
   if (!argType || argType->hasUnresolvedType() || argType->hasTypeVariable() ||
       argType->hasTypeParameter()) {
-    // Let's type check argument expression without any contextual information.
-    ConcreteDeclRef ref = nullptr;
-    auto typeResult =
-        TC.getTypeOfExpressionWithoutApplying(argExpr, CS.DC, ref);
-    if (!typeResult.hasValue())
+    auto *argTuple = dyn_cast<TupleExpr>(argExpr);
+    if (!argTuple) {
+      // Bail out if we don't have a well-formed argument list.
       return false;
+    }
+    
+    // Let's type check individual argument expressions without any
+    // contextual information to try to recover an argument type that
+    // matches what the user actually wrote instead of what the typechecker
+    // expects.
+    SmallVector<TupleTypeElt, 4> elts;
+    for (auto *el : argTuple->getElements()) {
+      ConcreteDeclRef ref = nullptr;
+      auto typeResult =
+        TC.getTypeOfExpressionWithoutApplying(el, CS.DC, ref);
```

Comment:
`getTypeOfExpressionWithoutApplying()` has returned an `Optional<Type>` since it was introduced in 69fdca43daabe84a84aa96ccea0e579922573d3c.